### PR TITLE
HPCC-31990 Add timeout to DNS lookups for soapcalls

### PR DESCRIPTION
**Pull request by mckellyln**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

**Comment by mckellyln** [comment]
@ghalliday initial thoughts ?

File: system/jlib/jsocket.cpp [hunk 1]
```diff
@@ -466,6 +467,200 @@ extern jlib_decl bool queryKeepAlive(int &time, int &intvl, int &probes)
...
+
+MODULE_INIT(INIT_PRIORITY_STANDARD)
+{
+    return true;
+}
+
+MODULE_EXIT()
+{
+    // BUG: this (and all other MODULE_EXITs) are not called for Thor and Roxie because they are
+    //      stopped via SIGTERM signal and thus exit() and the atexit handlers are not called ...
```
Comment (mckellyln) [inline system/jlib/jsocket.cpp:656]: Is this as designed/intended ?  I want to make sure a reaper thread is terminated before the process ends, but it seems MODULE_EXIT() is not called when Thor and Roxie end.
If I call exit() on SIGTERM then its called, but some other issues occur at process end that need to be looked into.

**Comment by mckellyln** [comment]
I think I'll change to using a threadpool and then no need for a reaper thread.

File: system/jlib/jsocket.cpp [hunk 2]
```diff
@@ -466,6 +467,200 @@ extern jlib_decl bool queryKeepAlive(int &time, int &intvl, int &probes)
...
+        bool ret = semait.wait(timeoutms);
+        if (ret && ended)
+        {
+            if (result)
+                memcpy(resAddr, netaddr, sizeof(netaddr));
+            if (join())
+                return true;
+        }
+        {
+            CriticalBlock block(queryDNSCS);
```
Comment (mckellyln) [inline system/jlib/jsocket.cpp:512]: Or a pthread_kill() here ???

Comment (mckellyln) [reply system/jlib/jsocket.cpp:512]: I think we could pthread_cancel() here to help stop thread asap but still would need to check for it / cleanup in the reaper thread.

**Review by ghalliday — COMMENTED** [review]
@mckellyln in general looks good.  A few minor suggestions to clean up code and questions about timeouts.  I didn't see any logic errors though.

**Comment by mckellyln** [comment]
switching to a thread pool

**Comment by mckellyln** [comment]
Closing this PR and will issue a new PR using a threadpool.

**Pull request closed without merging**